### PR TITLE
Add Tpm vendor prop, die_id to iMX

### PR DESCRIPTION
**Pull request by dmcilvaney**
Vendor properties are extra properties which can be added to a specific platform. In this case MSR added a "com.microsoft.ta.endorsementSeed" back in ~2015-16. This property generates a unique ID for the TA which is tied to the underlying hardware.

In the case of the original implementation it appears they were targeting by die ID which probably implies each family of devices would end up with the same endorsement seeds. For our use case we prefer to have unique seeds for every device.

To improve the quality of the keys generated `tee_otp_get_die_id()` was implemented for iMX. This consumes the second half of the BKEK which is left unused by `tee_otp_get_hw_unique_key()`. The weak implementation currently in use uses 'beef' repeated as needed.

This change **WILL break** any RPMB keys which have been generated from hardware ID and fused. Test keys should remain uneffected, although previously encrypted data may need to be cleared from RPMB since OP-TEE uses `tee_otp_get_die_id()` to generate a number of keys.

Signed-off-by: Daniel McIlvaney <damcilva@microsoft.com>

### CODE REVIEW

File: core/drivers/fsl_sec/hw_key_blob.c [hunk 1]
```diff
@@ -217,20 +224,57 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
...
+
+	res = init_and_roll_forward_master_key();
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to init normal priblob otpmk identity from CAAM");
+		return res;
+	}
+
+	// Use the first 16 bytes for the HUK, the last 16 for the die ID.
+	res = get_hw_unq_key_blob_hw(temp_buffer, KEY_BLOB_SIZE);
+
```
Comment (christopherco) [inline core/drivers/fsl_sec/hw_key_blob.c]: Do you need to handle a bad return value here too? You end up stomping it on the next call

File: core/drivers/fsl_sec/hw_key_blob.c [hunk 2]
```diff
@@ -217,20 +224,57 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
...
 	} else {
 		DMSG("Hardware Unique Key Retrieved:");
-		DHEXDUMP(&hwkey->data[0], HW_UNIQUE_KEY_LENGTH);
+	}
+
+	return res;
+}
+
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
```
Comment (neilsh-msft) [inline core/drivers/fsl_sec/hw_key_blob.c]: get_hw_unq_key_blob_hw uses int for size, should we be consistent?

File: core/drivers/fsl_sec/hw_key_blob.c [hunk 3]
```diff
@@ -217,20 +224,57 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
...
+	TEE_Result res;
+	uint8_t temp_buffer[KEY_BLOB_SIZE];
+	uint8_t *remaining_buffer = &(temp_buffer[HW_UNIQUE_KEY_LENGTH]);
+	uint32_t remaining_size = KEY_BLOB_SIZE - HW_UNIQUE_KEY_LENGTH;
+
+	// Make sure there is some data left available after taking the HUK.
+	COMPILE_TIME_ASSERT(KEY_BLOB_SIZE > HW_UNIQUE_KEY_LENGTH);
+
+	if (len > TEE_SHA256_HASH_SIZE) {
```
Comment (neilsh-msft) [inline core/drivers/fsl_sec/hw_key_blob.c]: Bad error check, should be if (len < TEE_SHA256_HASH_SIZE) ?

File: core/drivers/fsl_sec/hw_key_blob.c (hunk 2 above)
Comment (dmcilvaney) [reply core/drivers/fsl_sec/hw_key_blob.c]: Its implementing a weak function (which uses int for some reason). Probably just old OP-TEE code no one has looked at in a while.

File: core/drivers/fsl_sec/hw_key_blob.c [hunk 4]
```diff
@@ -217,20 +224,57 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
...
 	res = init_and_roll_forward_master_key();
 	if (res != TEE_SUCCESS) {
 		EMSG("Failed to init normal priblob otpmk identity from CAAM");
 		return res;
 	}
 
-	res = get_hw_unq_key_blob_hw(&hwkey->data[0], HW_UNIQUE_KEY_LENGTH);
+	// Use the first 16 bytes for the HUK, the last 16 for the die ID.
+	res = get_hw_unq_key_blob_hw(temp_buffer, KEY_BLOB_SIZE);
+	memcpy(&hwkey->data[0], temp_buffer, HW_UNIQUE_KEY_LENGTH);
```
[the remark this answers is not in the document]
Comment (dmcilvaney) [reply core/drivers/fsl_sec/hw_key_blob.c]: Good call, will do.

File: core/drivers/fsl_sec/hw_key_blob.c (hunk 3 above)
Comment (dmcilvaney) [reply core/drivers/fsl_sec/hw_key_blob.c]: This API is very odd, you can request as much data as you want from it but there is a fixed amount of data to return (at least when reading from fuses). OP-TEE maintainers feel SHA256 is the largest this value should ever be (from issue I opened a while ago), so I'm following that. Requesting more than the available data is an error in this case.

tee_hash_createdigest() will truncate the hash to fit the buffer length provided (and fail if the buffer is longer)

File: core/drivers/fsl_sec/hw_key_blob.c (hunk 1 above)
Comment (dmcilvaney) [reply core/drivers/fsl_sec/hw_key_blob.c]: Good catch, missed that one.

File: core/drivers/fsl_sec/hw_key_blob.c [hunk 5]
```diff
@@ -216,22 +222,69 @@ TEE_Result get_hw_unq_key_blob_hw(uint8_t *hw_key, int size)
...
+		goto out;
+
+	res = tee_hash_createdigest(TEE_ALG_SHA256, remaining_buffer, remaining_size,
+				buffer, len);
+
+out:
+	if (res != TEE_SUCCESS)
+		EMSG("Die ID Failed");
+	else
+		DMSG("Die ID Retrieved:");
```
Comment (neilsh-msft) [inline core/drivers/fsl_sec/hw_key_blob.c]: nit: any reason Failed and Retrieved are capitalized?  Also the colon after retrieved suggests that something follows it, but there's not extra print statement.

Comment (dmcilvaney) [reply core/drivers/fsl_sec/hw_key_blob.c]: We used to dump the (super secret) value of the buffer. I removed that but missed the colon. I'll update it when I add the testing fallback which returns a zero buffer.

**Comment by dmcilvaney** [comment]
I've added a fall back mechanism which will return a zeroed out buffer if the CAAM unique key is unavailable AND the RPMB test key is in use. This occurs when HAB is not turned on but CFG_FSL_SEC=y is used.
